### PR TITLE
JSUI-2951 Retrieve All facets from the SearchInterface directly instead of looking inside the root element

### PR DESCRIPTION
**Pull request by ThibodeauJF**
https://coveord.atlassian.net/browse/JSUI-2951

An improvement upon this PR: https://github.com/coveo/search-ui/pull/1157
The issue is with external components, when they are registered with the SearchInterface but not inside the Dom element.

Question: should I go forward and replace methods that were calling `ComponentsTypes.getAllFacetsInstance` with the goal of getting instances from the root of the SearchInterface?
Should I change the name `ComponentsTypes.getAllFacetsInstance` to something like `ComponentsTypes.getAllFacetsInstanceInElement`?

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/utils/ComponentsTypes.ts [hunk 1]
```diff
@@ -1,6 +1,10 @@
 import { Component } from '../ui/Base/Component';
 import { $$, Dom } from './Dom';
 
+export interface IComponentsTypesSearchInterface {
```
Comment (ThibodeauJF) [inline src/utils/ComponentsTypes.ts:4]: Created that to prevent further circular dependencies

**Comment by samisayegh** [comment]
> Question: should I go forward and replace methods that were calling ComponentsTypes.getAllFacetsInstance with the goal of getting instances from the root of the SearchInterface?

I think it makes sense to do so, to have a consistent behaviour. That is, only facets inside the SearchInterface are in-scope. I think the risk of a breaking change is low?

> Should I change the name ComponentsTypes.getAllFacetsInstance to something like ComponentsTypes.getAllFacetsInstanceInElement?

For this one, I think the final name you used is `getAllFacetsFromSearchInterface` which I find clear explains what the function is doing,

File: src/utils/ComponentsTypes.ts [hunk 2]
```diff
@@ -32,4 +36,11 @@ export class ComponentsTypes {
   public static getAllFacetsInstance(root: HTMLElement | Dom) {
     return ComponentsTypes.getAllFacetsElements(root).map(element => Component.get(element) as Component);
   }
+
+  public static getAllFacetsFromSearchInterface(searchInterface: IComponentsTypesSearchInterface) {
+    return ComponentsTypes.allFacetsType.reduce(
+      (facets: Component[], facetType: string) => facets.concat(searchInterface.getComponents(facetType)),
+      []
+    );
```
Comment (btaillon) [inline src/utils/ComponentsTypes.ts:44]: Would that be equivalent to the following?
```ts
return _.flatten(
  ComponentsTypes.allFacetsType.map(facetType => searchInterface.getComponents(facetType))
);
```
If so, I think a map and flatten would be easier to read.

Comment (ThibodeauJF) [reply src/utils/ComponentsTypes.ts:44]: I don't think the reduce() that complicated (or more complicated than the flatten), it's a bit more optimized and I don't need to type the return value. 
I could also just do a forEach() and push to an array.

**Pull request merged**